src/bandersnatch/master.py

```diff
@@ -141,58 +136,47 @@ async def url_fetch(
                     fd.write(chunk)
 
     @property
-    def xmlrpc_url(self) -> str:
-        return f"{self.url}/pypi"
-
-    # TODO: Potentially make USER_AGENT more accessible from aiohttp-xmlrpc
-    async def _gen_custom_headers(self) -> dict[str, str]:
-        # Create dummy client so we can copy the USER_AGENT + prepend bandersnatch info
-        dummy_client = ServerProxy(self.xmlrpc_url, loop=self.loop)
-        custom_headers = {
-            "User-Agent": (
-                f"bandersnatch {bandersnatch.__version__} {dummy_client.USER_AGENT}"
-            )
+    def simple_url(self) -> str:
+        return f"{self.url}/simple/"
+
+    # TODO: Potentially make USER_AGENT more accessible from aiohttp
+    @property
+    def _custom_headers(self) -> dict[str, str]:
+        return {
+            "User-Agent": f"bandersnatch {bandersnatch.__version__}",
+            # the simple API use headers to return JSON
+            "Accept": "application/vnd.pypi.simple.v1+json",
         }
-        await dummy_client.close()
-        return custom_headers
-
-    async def _gen_xmlrpc_client(self) -> ServerProxy:
-        custom_headers = await self._gen_custom_headers()
-        client = ServerProxy(
-            self.xmlrpc_url,
-            client=self.session,
-            loop=self.loop,
-            headers=custom_headers,
-        )
-        return client
 
-    # TODO: Add an async context manager to aiohttp-xmlrpc to replace this function
-    async def rpc(self, method_name: str, serial: int = 0) -> Any:
-        try:
-            client = await self._gen_xmlrpc_client()
-            method = getattr(client, method_name)
-            if serial:
-                return await method(serial)
-            return await method()
-        except TimeoutError as te:
-            logger.error(f"Call to {method_name} @ {self.xmlrpc_url} timed out: {te}")
-
-    async def all_packages(self) -> Any:
-        all_packages_with_serial = await self.rpc("list_packages_with_serial")
-        if not all_packages_with_serial:
-            raise XmlRpcError("Unable to get full list of packages")
-        return all_packages_with_serial
+    async def fetch_simple_index(self) -> Any:
+        """Return a mapping of all project data from the PyPI Index API"""
+        logger.debug(f"Fetching simple JSON index from {self.simple_url}")
+        async with self.session.get(
+            self.simple_url, headers=self._custom_headers
+        ) as response:
+            simple_index = await response.json()
```

Comment:
There is no guarantee that the return is actually JSON according to the spec https://peps.python.org/pep-0691/#content-types.

Reply:
Good call, we did that more for migration and writing defensive code. So we should wrap this and error saying didn't get JSON when we asked for it so the error is clear ...